include/zephyr/mgmt/ec_host_cmd/backend.h

```diff
@@ -0,0 +1,149 @@
+	/** Buffer to hold received data. The buffer is provided by the handler if
+	 * EC_HOST_CMD_HANDLER_RX_BUFFER > 0. Otherwise, the backend should provide
+	 * the buffer on its own and overwrites @a buf pointer in the init function.
+	 */
+	uint8_t *buf;
+	/** Number of bytes written to @a buf by backend. */
+	size_t len;
+	/** The backend gives @a handler_owns, when data in @a buf are ready.
+	 * The handler takes @a handler_owns to read data in @a buf.
+	 */
+	struct k_sem handler_owns;
+};
+
+/**
+ * @brief Context for host command backend and handler to pass tx data
+ */
+struct ec_host_cmd_tx_buf {
+	/** Data to write to the host The buffer is provided by the handler if
+	 * EC_HOST_CMD_HANDLER_TX_BUFFER > 0. Otherwise, the backend should provide
+	 * the buffer on its own and overwrites @a buf pointer and @a len_max
+	 * in the init function.
+	 */
+	void *buf;
+	/** Number of bytes to write from @a buf */
+	size_t len;
+	/** Size of @a buf */
+	size_t len_max;
+};
+
+/**
+ * @brief Config for the eSPI backend
+ */
+struct ec_host_cmd_espi_config {
+	/* eSPI device instance */
+	const struct device *dev;
+};
+
+/**
+ * @brief Config for the SHI backend
+ */
+struct ec_host_cmd_shi_ite_config {
+	/* Device instance */
+	const struct device *dev;
+};
+
+/**
+ * @brief Config for the SHI backend
+ */
+struct ec_host_cmd_shi_npcx_config {
+	/* Device instance */
+	const struct device *dev;
+};
+
+/**
+ * @brief Initialize a host command backend
+ *
+ * This routine initializes a host command backend. It includes initialization
+ * a device used to communication and setting up buffers.
+ *
+ * @param backend [in] A pointer to the backend structure for the driver instance.
+ * @param config  [in] A pointer to backend dependent config data
+ * @param rx_ctx  [in,out] A pointer to the receive context object.
+ *		  These objects are used to receive data from the driver
+ *		  when the host send data. The buf member can be assigne by the
+ *		  backend.
+ * @param tx      [in,out] A pointer to the transmit buffer object. The buf and
+ *		  len_max members can be assigne by the backend. These objects
+ *		  are used to send data by the backend with the
+ *		  ec_host_cmd_backend_api_send function.
+ *
+ * @retval 0 if successful
+ */
+typedef int (*ec_host_cmd_backend_api_init)(const struct ec_host_cmd_backend *backend,
+					    const void *config, struct ec_host_cmd_rx_ctx *rx_ctx,
+					    struct ec_host_cmd_tx_buf *tx);
+
+/**
+ * @brief Sends data to the host
+ *
+ * Sends data from tx buf that was passed via ec_host_cmd_backend_api_init
+ * function.
+ *
+ * @param backend Pointer to the backed to send data
+ *
+ * @retval 0 if successful
+ */
+typedef int (*ec_host_cmd_backend_api_send)(const struct ec_host_cmd_backend *backend);
+
+__subsystem struct ec_host_cmd_backend_api {
+	ec_host_cmd_backend_api_init init;
+	ec_host_cmd_backend_api_send send;
+};
+
+struct ec_host_cmd_backend *ec_host_cmd_backend_get_espi(void);
```

Comment:
I don't get why it removes the 'generic' aspect of the backends. This function only returns a pointer to espi backend to pass it to the `ec_host_cmd_init` function.

Reply:
Because you have to implement it for every backend available, and adding new backend will require to add similar function for this backend. If you would have a function "ec_host_cmd_backend_get", that takes some pointer to device, as all APIs are doing, then you would have a generic function, independent of any backend.

Reply:
If the `zephyr_host_cmd_backend` is not chosen, how would you know which backend is connected to this device. You need to call `ec_host_cmd_init` first to pass the device pointer as config to a backed. The backend pointer is required to call the `ec_host_cmd_init`, so it is a loop. You need a way to get pointer to a backend. The backends don't have their nodes dts, so there is no separated devices structures, with their API. The devices have peripheral API, not the backend API. The backend API is not available for application code. Anyway, this approach is common in Zephyr e.g. for [shell](https://github.com/zephyrproject-rtos/zephyr/blob/22ff66f2401e57e887322dacbe0d5d5f5f4322aa/include/zephyr/shell/shell_uart.h).